onnxruntime/core/providers/cpu/math/matmul_helper.h

```diff
@@ -22,20 +23,25 @@ class MatMulComputeHelper {
     size_t right_num_dims = right_shape.NumDimensions();
     ORT_RETURN_IF_NOT(left_num_dims >= 1 && right_num_dims >= 1);
 
-    // special case for right_shape being 2D and left_shape > 2D by flattening left_shape to 2D
-    // note that padding 1s in front of the right shape can be flattened too
-    if (left_num_dims >= 2 && right_num_dims >= 2 &&
+    // Special cases below for right_shape being 2D and left_shape > 2D by flattening left_shape to 2D
+    // Note that padding 1s in front of the right_shape can be flattened too
+    // A: [M1, M2, ... K], B: [K, N]
+    // A: [M1, M2, ... K], B: [N, K]^T
+    // A: [M1, M2, ... K], B: [1, ..., 1, K, N]
+    // A: [M1, M2, ... K], B: [1, ..., 1, N, K]^T
+    if (!transa && left_num_dims >= 2 && right_num_dims >= 2 &&
         right_shape.SizeToDimension(right_num_dims - 1) == right_shape[right_num_dims - 2]) {
```

Comment:
Should there be new unit tests to validate the changes here?

Reply:
yes, there's transpose_malmul tests in training branch covered. In the current logic, the new params will be false and the logic remains the same for master. We can consider to add test cases in future release.